src/test/rpc/Subscribe_test.cpp

```diff
@@ -382,13 +383,269 @@ class Subscribe_test : public beast::unit_test::suite
         BEAST_EXPECT(jv[jss::status] == "success");
     }
 
+    void
+    testSubByUrl()
+    {
+        using namespace jtx;
+        testcase("Subscribe by url");
+        Env env {*this};
+
+        Json::Value jv;
+        jv[jss::url] = "http://localhost/events";
+        jv[jss::url_username] = "admin";
+        jv[jss::url_password] = "password";
+        jv[jss::streams] = Json::arrayValue;
+        jv[jss::streams][0u] = "validations";
+        auto jr = env.rpc("json", "subscribe", to_string(jv)) [jss::result];
+        BEAST_EXPECT(jr[jss::status] == "success");
+
+        jv[jss::streams][0u] = "ledger";
+        jr = env.rpc("json", "subscribe", to_string(jv)) [jss::result];
+        BEAST_EXPECT(jr[jss::status] == "success");
+
+        jr = env.rpc("json", "unsubscribe", to_string(jv)) [jss::result];
+        BEAST_EXPECT(jr[jss::status] == "success");
+
+        jv[jss::streams][0u] = "validations";
+        jr = env.rpc("json", "unsubscribe", to_string(jv)) [jss::result];
+        BEAST_EXPECT(jr[jss::status] == "success");
+    }
+
+    void
+    testSubErrors(bool subscribe)
+    {
+        using namespace jtx;
+        auto const method = subscribe ? "subscribe" : "unsubscribe";
+        testcase << "Error cases for " << method;
+
+        Env env {*this};
+        auto wsc = makeWSClient(env.app().config());
+
+        {
+            auto jr = env.rpc("json", method, "{}") [jss::result];
+            BEAST_EXPECT(jr[jss::error] == "invalidParams");
+            BEAST_EXPECT(jr[jss::error_message] == "Invalid parameters.");
+        }
+
+        {
+            Json::Value jv;
+            jv[jss::url] = "not-a-url";
+            jv[jss::username] = "admin";
+            jv[jss::password] = "password";
+            auto jr = env.rpc("json", method, to_string(jv)) [jss::result];
+            if (subscribe)
+            {
+                BEAST_EXPECT(jr[jss::error] == "internal");
+                BEAST_EXPECT(jr[jss::error_message] == "Internal error.");
```

Comment:
Nice test.  Seems like we should have a guideline that a user should not be able to (easily) do something that returns an `internal error`.  Would it be hard for the error handling mechanisms to trap this case and handle it better?  This comment applies to the next test case as well.  Not insisting, just wondering...